File: netlify/functions/script_detail.js
const https = require('https');
const sqlite3 = require('sqlite3');
const fs = require('fs');
const path = require('path');
const os = require('os');

// Dropbox直接ダウンロードURL
const DROPBOX_URL = 'https://www.dropbox.com/scl/fi/dljhp6xzshdgvq7vqk3sz/sunsun_final_dialogue_database_proper.db?rlkey=qlf38ydm1b0n0ocsdbpjx0ih8&st=2h1nmfhq&dl=1';

let dbPath = null;

function downloadDatabase() {
  return new Promise((resolve, reject) => {
    if (dbPath && fs.existsSync(dbPath)) {
      // ファイルサイズをチェック
      const stats = fs.statSync(dbPath);
      if (stats.size > 1000000) { // 1MB以上なら有効とみなす
        resolve(dbPath);
        return;
      }
    }

    dbPath = path.join(os.tmpdir(), `sunsun_database_${Date.now()}.db`);
    console.log('Downloading database to:', dbPath);
    
    const file = fs.createWriteStream(dbPath);
    
    // User-Agentを追加してブラウザっぽくする
    const options = {
      headers: {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36'
      }
    };
    
    const request = https.get(DROPBOX_URL, options, (response) => {
      console.log('Response status:', response.statusCode);
      
      // リダイレクトをフォロー（複数回対応）
      if (response.statusCode >= 300 && response.statusCode < 400 && response.headers.location) {
        console.log('Redirected to:', response.headers.location);
        
        const redirectUrl = response.headers.location.startsWith('http') 
          ? response.headers.location 
          : 'https:' + response.headers.location;
        
        const redirectRequest = https.get(redirectUrl, options, (redirectResponse) => {
          console.log('Redirect response status:', redirectResponse.statusCode);
          
          if (redirectResponse.statusCode >= 300 && redirectResponse.statusCode < 400 && redirectResponse.headers.location) {
            // 2回目のリダイレクト
            const finalUrl = redirectResponse.headers.location.startsWith('http') 
              ? redirectResponse.headers.location 
              : 'https:' + redirectResponse.headers.location;
            
            console.log('Final redirect to:', finalUrl);
            
            const finalRequest = https.get(finalUrl, options, (finalResponse) => {
              handleResponse(finalResponse, file, dbPath, resolve, reject);
            });
            
            finalRequest.on('error', reject);
          } else {
            handleResponse(redirectResponse, file, dbPath, resolve, reject);
          }
        });
        
        redirectRequest.on('error', reject);
      } else {
        handleResponse(response, file, dbPath, resolve, reject);
      }
    });
    
    request.on('error', reject);
    file.on('error', reject);
  });
}

function handleResponse(response, file, dbPath, resolve, reject) {
  let downloadedBytes = 0;
  
  response.on('data', (chunk) => {
    downloadedBytes += chunk.length;
  });
  
  response.pipe(file);
  
  file.on('finish', () => {
    file.close();
    
    console.log('Download finished, size:', downloadedBytes);
    
    // ファイルサイズをチェック
    const stats = fs.statSync(dbPath);
    console.log('File size on disk:', stats.size);
    
    if (stats.size < 1000000) { // 1MB未満は問題
      const errorMsg = `Downloaded file too small: ${stats.size} bytes`;
      console.error(errorMsg);
      reject(new Error(errorMsg));
      return;
    }
    
    // SQLite ファイルかチェック
    const buffer = fs.readFileSync(dbPath, { start: 0, end: 15 });
    const header = buffer.toString();
    
    if (!header.includes('SQLite format 3')) {
      const errorMsg = `Not a valid SQLite file. Header: ${header}`;
      console.error(errorMsg);
      reject(new Error(errorMsg));
      return;
    }
    
    console.log('Valid SQLite database downloaded');
    resolve(dbPath);
  });
}

exports.handler = async (event, context) => {
  const headers = {
    'Access-Control-Allow-Origin': '*',
    'Access-Control-Allow-Methods': 'GET, POST, OPTIONS',
    'Access-Control-Allow-Headers': 'Content-Type',
    'Content-Type': 'application/json'
  };

  // Handle OPTIONS request for CORS
  if (event.httpMethod === 'OPTIONS') {
    return {
      statusCode: 200,
      headers,
      body: ''
    };
  }

  // Handle GET request
  if (event.httpMethod === 'GET') {
    try {
      // URLパラメータを取得
      const queryStringParameters = event.queryStringParameters || {};
      const scriptName = queryStringParameters.script_name || '';
      const keyword = queryStringParameters.keyword || '';
      
      if (!scriptName) {
        return {
          statusCode: 400,
          headers,
          body: JSON.stringify({
            success: false,
            error: '台本名が必要です'
          })
        };
      }

      console.log('Script detail request for:', scriptName);
      console.log('Script name length:', scriptName.length);
      console.log('Script name hex:', Buffer.from(scriptName, 'utf8').toString('hex'));

      // データベースをダウンロード
      const dbFile = await downloadDatabase();
      
      return new Promise((resolve, reject) => {
        const db = new sqlite3.Database(dbFile);
        
        // 台本詳細情報を取得
        const scriptQuery = `
          SELECT DISTINCT
            script_name,
            script_url,
            release_date,
            youtube_title,
            youtube_url,
            youtube_video_id,
            themes,
            subjects,
            category
          FROM dialogues 
          WHERE script_name = ?
          LIMIT 1
        `;
        
        db.get(scriptQuery, [scriptName], (err, scriptInfo) => {
          if (err) {
            console.error('Script info query error:', err);
            db.close();
            reject(err);
            return;
          }
          
          if (!scriptInfo) {
            console.log('Script not found. Checking similar names...');
            
            // デバッグ: 似た名前を検索
            const debugQuery = "SELECT DISTINCT script_name FROM dialogues WHERE script_name LIKE '%B2231%' LIMIT 5";
            db.all(debugQuery, (debugErr, debugRows) => {
              console.log('Similar script names found:', debugRows);
              
              db.close();
              resolve({
                statusCode: 404,
                headers,
                body: JSON.stringify({
                  success: false,
                  error: '台本が見つかりません',
                  debug: {
                    searched_for: scriptName,
                    searched_length: scriptName.length,
                    similar_names: debugRows || []
                  }
                })
              });
            });
            return;
          }
          
          // セリフ詳細を取得
          let dialogueQuery;
          let params;
          
          if (keyword) {
            // キーワード指定時は該当セリフのみ
            dialogueQuery = `
              SELECT 
                character,
                dialogue,
                row_number
              FROM dialogues 
              WHERE script_name = ?
              AND dialogue LIKE ?
              AND dialogue IS NOT NULL 
              AND dialogue != ""
              ORDER BY row_number
            `;
            params = [scriptName, `%${keyword}%`];
          } else {
            // 全セリフ取得
            dialogueQuery = `
              SELECT 
                character,
                dialogue,
                row_number
              FROM dialogues 
              WHERE script_name = ?
              AND dialogue IS NOT NULL 
              AND dialogue != ""
              ORDER BY row_number
            `;
            params = [scriptName];
          }
          
          db.all(dialogueQuery, params, (err, rows) => {
            db.close();
            
            if (err) {
              console.error('Dialogue query error:', err);
              reject(err);
              return;
            }
            
            const dialogues = [];
            let matchCount = 0;
            
            rows.forEach(row => {
              const dialogueText = row.dialogue || '';
              let isMatch = false;
              
              if (keyword && dialogueText.toLowerCase().includes(keyword.toLowerCase())) {
                isMatch = true;
                matchCount++;
              }
              
              dialogues.push({
                character: row.character || '',
                dialogue: dialogueText,
                row_number: row.row_number || 0,
                is_match: isMatch
              });
            });
            
            // マッチ度計算（キーワード指定時のみ）
            const matchConfidence = keyword && dialogues.length > 0 
              ? matchCount / dialogues.length 
              : 0;
            
            const response = {
              success: true,
              data: {
                script_name: scriptInfo.script_name,
                script_url: scriptInfo.script_url || '',
                release_date: scriptInfo.release_date || '',
                youtube_title: scriptInfo.youtube_title || '',
                youtube_url: scriptInfo.youtube_url || '',
                youtube_video_id: scriptInfo.youtube_video_id || '',
                themes: scriptInfo.themes || '',
                subjects: scriptInfo.subjects || '',
                category: scriptInfo.category || '',
                total_dialogues: dialogues.length,
                match_count: matchCount,
                match_confidence: matchConfidence,
                keyword: keyword,
                dialogues: dialogues
              }
            };
            
            resolve({
              statusCode: 200,
              headers,
              body: JSON.stringify(response)
            });
          });
        });
      });
      
    } catch (error) {
      console.error('Error in script detail handler:', error);
      
      return {
        statusCode: 500,
        headers,
        body: JSON.stringify({
          success: false,
          error: error.message
        })
      };
    }
  }

  // Method not allowed
  return {
    statusCode: 405,
    headers,
    body: JSON.stringify({ error: 'Method not allowed' })
  };
};

File: netlify/functions/script_detail.py
import json
import sqlite3
import urllib.request
import tempfile
import os
import ssl

# Dropbox直接ダウンロードURL
DROPBOX_URL = 'https://www.dropbox.com/scl/fi/dljhp6xzshdgvq7vqk3sz/sunsun_final_dialogue_database_proper.db?rlkey=qlf38ydm1b0n0ocsdbpjx0ih8&st=2h1nmfhq&dl=1'

# データベース一時ファイル
db_path = None

def download_database():
    """Dropboxからデータベースファイルをダウンロード"""
    global db_path
    
    if db_path and os.path.exists(db_path):
        return db_path
    
    try:
        # 一時ファイルを作成
        temp_fd, db_path = tempfile.mkstemp(suffix='.db')
        os.close(temp_fd)
        
        # SSL証明書検証をスキップ
        ssl_context = ssl.create_default_context()
        ssl_context.check_hostname = False
        ssl_context.verify_mode = ssl.CERT_NONE
        
        opener = urllib.request.build_opener(urllib.request.HTTPSHandler(context=ssl_context))
        urllib.request.install_opener(opener)
        
        urllib.request.urlretrieve(DROPBOX_URL, db_path)
        
        return db_path
        
    except Exception as e:
        print(f"Error downloading database: {e}")
        raise

def get_db_connection():
    """データベース接続を取得"""
    db_file = download_database()
    conn = sqlite3.connect(db_file)
    conn.row_factory = sqlite3.Row
    return conn

def main(event, context):
    """Netlify Function handler"""
    
    headers = {
        'Access-Control-Allow-Origin': '*',
        'Access-Control-Allow-Methods': 'GET, POST, OPTIONS',
        'Access-Control-Allow-Headers': 'Content-Type',
        'Content-Type': 'application/json'
    }
    
    # Handle OPTIONS request for CORS
    if event['httpMethod'] == 'OPTIONS':
        return {
            'statusCode': 200,
            'headers': headers,
            'body': ''
        }
    
    # Handle GET request
    if event['httpMethod'] == 'GET':
        try:
            # URLパラメータを取得
            query_params = event.get('queryStringParameters', {}) or {}
            script_name = query_params.get('script_name', '').strip()
            keyword = query_params.get('keyword', '').strip()
            
            if not script_name:
                response = {
                    'success': False,
                    'error': '台本名が必要です'
                }
                return {
                    'statusCode': 400,
                    'headers': headers,
                    'body': json.dumps(response)
                }
            
            # データベース検索
            conn = get_db_connection()
            cursor = conn.cursor()
            
            # 台本詳細情報を取得
            script_query = '''
                SELECT DISTINCT
                    script_name,
                    script_url,
                    release_date,
                    youtube_title,
                    youtube_url,
                    youtube_video_id,
                    themes,
                    subjects,
                    category
                FROM dialogues 
                WHERE script_name = ?
                LIMIT 1
            '''
            
            cursor.execute(script_query, (script_name,))
            script_info = cursor.fetchone()
            
            if not script_info:
                response = {
                    'success': False,
                    'error': '台本が見つかりません'
                }
                return {
                    'statusCode': 404,
                    'headers': headers,
                    'body': json.dumps(response)
                }
            
            # セリフ詳細を取得
            if keyword:
                # キーワード指定時は該当セリフのみ
                dialogue_query = '''
                    SELECT 
                        character,
                        dialogue,
                        row_number
                    FROM dialogues 
                    WHERE script_name = ?
                    AND dialogue LIKE ?
                    AND dialogue IS NOT NULL 
                    AND dialogue != ""
                    ORDER BY row_number
                '''
                cursor.execute(dialogue_query, (script_name, f'%{keyword}%'))
            else:
                # 全セリフ取得
                dialogue_query = '''
                    SELECT 
                        character,
                        dialogue,
                        row_number
                    FROM dialogues 
                    WHERE script_name = ?
                    AND dialogue IS NOT NULL 
                    AND dialogue != ""
                    ORDER BY row_number
                '''
                cursor.execute(dialogue_query, (script_name,))
            
            dialogues = []
            match_count = 0
            
            for row in cursor.fetchall():
                dialogue_text = row['dialogue'] or ''
                is_match = False
                
                if keyword and keyword.lower() in dialogue_text.lower():
                    is_match = True
                    match_count += 1
                
                dialogues.append({
                    'character': row['character'] or '',
                    'dialogue': dialogue_text,
                    'row_number': row['row_number'] or 0,
                    'is_match': is_match
                })
            
            # マッチ度計算（キーワード指定時のみ）
            match_confidence = 0
            if keyword and dialogues:
                match_confidence = match_count / len(dialogues)
            
            conn.close()
            
            response = {
                'success': True,
                'data': {
                    'script_name': script_info['script_name'],
                    'script_url': script_info['script_url'] or '',
                    'release_date': script_info['release_date'] or '',
                    'youtube_title': script_info['youtube_title'] or '',
                    'youtube_url': script_info['youtube_url'] or '',
                    'youtube_video_id': script_info['youtube_video_id'] or '',
                    'themes': script_info['themes'] or '',
                    'subjects': script_info['subjects'] or '',
                    'category': script_info['category'] or '',
                    'total_dialogues': len(dialogues),
                    'match_count': match_count,
                    'match_confidence': match_confidence,
                    'keyword': keyword,
                    'dialogues': dialogues
                }
            }
            
            return {
                'statusCode': 200,
                'headers': headers,
                'body': json.dumps(response)
            }
            
        except Exception as e:
            print(f"Error in script detail handler: {e}")
            response = {
                'success': False,
                'error': str(e)
            }
            
            return {
                'statusCode': 500,
                'headers': headers,
                'body': json.dumps(response)
            }
    
    # Method not allowed
    return {
        'statusCode': 405,
        'headers': headers,
        'body': json.dumps({'error': 'Method not allowed'})
    }